### ISSUE TITLE
Garage Door Opener

### ISSUE THREAD

**Issue by fubano** [ikalchev/HAP-python#5]
Thanks for sharing this project, I find python a little easier to using NodeJS. I can get the TempSensor, LightBulb to work on my Pi Zero W, but my attempt to mash code from multiple existing accessories to create a Garage Door Opener fails (it pairs and then unpairs right away). Can someone can point to the issues with this code? Not sure __getstate__ is needed at all, my suspicion was that the required characteristic state is not getting set correctly.

class GarageDoorOpener(Accessory):
    category = Category.GARAGE_DOOR_OPENER

    @classmethod
    def _gpio_setup(_cls, pulse_pin, sense_pin):
        GPIO.setmode(GPIO.BCM)
        GPIO.setup(pulse_pin, GPIO.OUT)
        GPIO.setup(sense_pin, GPIO.IN)
        GPIO.add_event_detect(sense_pin, GPIO.BOTH, callback=self._detected)

    def __init__(self, *args, pulse_pin=26, sense_pin=22, **kwargs):
        super(GarageDoorOpener, self).__init__(*args, **kwargs)

        self.CurrentDoorState = self.get_service("GarageDoorOpener").get_characteristic("CurrentDoorState")
        self.ObstructionDetected = self.get_service("GarageDoorOpener").get_characteristic("ObstructionDetected")

        self.pulse_pin = pulse_pin
        self.sense_pin = sense_pin
        self._gpio_setup(pulse_pin,sense_pin)

    def __getstate__(self):
        state = super(GarageDoorOpener, self).__getstate__()
        return state

    def __setstate__(self, state):
        self.__dict__.update(state)
        self._gpio_setup(self.pulse_pin, self.sense_pin)

    def _pulse_door(self, value):
        GPIO.output(self.pulse_pin, GPIO.HIGH)
        time.sleep(2)
        GPIO.output(self.pulse_pin, GPIO.LOW)
        time.sleep(2)

    def _set_services(self):
        super(GarageDoorOpener, self)._set_services()

        garage_door_opener_service = loader.get_serv_loader().get("GarageDoorOpener")
        self.add_service(garage_door_opener_service)
        garage_door_opener_service.get_characteristic("TargetDoorState").setter_callback = self._pulse_door

    def _detected(self, _pin):
        if(GPIO.input(self.sense_pin)):
            self.CurrentDoorState.set_value(1)
        else:
            self.CurrentDoorState.set_value(0)

    def stop(self):
        super(GarageDoorOpener, self).stop()
        GPIO.cleanup()
~

**Comment by rodriguez-lucas** [comment]
I didn't read the code you wrote, but I had the same problem with the door service: the device pairs and then it unpairs. 
As you suspected there is a problem with `setstate`/`getstate` (I don't remember which one was in my case). If you look at `CurrentDoorState` characteristic ([characteristics.json](../blob/master/pyhap/resources/characteristics.json) line 238), the permissions are `pr`, `ev` and `null`. I don't know why but `null` is for `uncnotify` according to [gen_hap_types.py](../blob/master/gen_hap_types.py) (I don't have an apple developer id to verify some things, or maybe there is another source to check them which I don't know). 

Anyway, removing the `null` permission from the characteristic/s you're using solves your problem.
You should remove all `null`s from all the characteristics that you want to use. I remove them all for simplicity.
For example if you want to use `CurrentDoorState`, you should replace it in [characteristics.json](../blob/master/pyhap/resources/characteristics.json):

```json
  "CurrentDoorState": {
      "Format": "uint8",
      "UUID": "0000000E-0000-1000-8000-0026BB765291",
      "Permissions": [
         "pr",
         "ev"
      ],
      "ValidValues": {
         "Closing": 3,
         "Stopped": 4,
         "Open": 0,
         "Closed": 1,
         "Opening": 2
      }
   }
```
**Discalimer**: I don't know why the `uncnotify` is represented by `null` (json) / `None` (python), and I don't know the effects of removing this permission. With that said, the device is working fine for me.

**Comment by ikalchev** [comment]
Hi,

Thanks for reporting and investigating this! I will have a look and try to understand and fix the problem with uncnotify in the following days. To be honest, I haven’t read Apples specs, so chances are I messed up this bit (specs weren’t available for non-MFI members when I implemented this).

Will keep you posted :)

**Comment by ikalchev** [comment]
Okay, pushed a quick fix - I removed the null permission until I find the correct representation of it. I tested with a fake garage door opener and it seems to work. Please let me know if the problem continues :)

@rodriguez-lucas, @fubano  What is the problem with set/get state, I couldn't reproduce any problem involving this? A little clarification into what may be going on:

When the accessory pairs, it preserves its state in a file `accessory.pickle`(this is to preserve the state of the accessories, paired clients, char values, etc.). When you restart the accessory, it restores the object from this file. This means that the `init` method of the Accessory is not called (the `setstate` method is called instead), i.e. GPIO setup code is not called. In other words, in `setstate` you will need to configure GPIO as shown in the first comment. I hope this helps.

Best,
Ivan

**Comment by rodriguez-lucas** [comment]
Hi @ikalchev,

Sorry, I totally messed up with the names, the problem wasn't with `setstate`/`getstate`, instead was with characteristics methods: `set_value`/`get_value`. My hypotesis is that after pairing successfully, the phone app (e.g. Home) is requiring the state of the device, so `get_value` is called and there is some conflict with the `null` permission.

Again sorry for my mistake. I read `getstate` and assumed that it was the device state and not the **pickle** function...

**Comment by ikalchev** [comment]
Hey, no worries. You are right though, the null was indeed the problem. The change from today should fix this, could you please try and confirm? :)

**Comment by fubano** [comment]
Thanks to both @rodriguez-lucas and @ikalchev  for your input! I can now pair the accessory and get it to work (mostly). There is now an issue with the CurrentState reporting it gets stuck at "opening.." or "closing.." on the phone, though it executes the command to open/close correctly even with the status stuck. Looking into it.

Also, thanks to @ikalchev for the details about the pickle functions. I hope to run this on a read-only filesystem eventually for power failure tolerance and I need to figure out a way to incorporate this file write/read.

**Comment by ikalchev** [comment]
@fubano interesting, I haven't thought about read-only use cases. The file is created in [AccessoryDriver](pyhap/accessory_driver.py), if that helps. To clarify the use of the pickle, the main point is to preserve the public key of the paired client, which is received during the pairing, so I don't see how this can be removed.

You can try to overwrite the `persist` method of the driver (which is where all the writing happens) and instead read/write the pickle to a remote server on another machine in your home, which has tolerance to power failures. I am open to suggestion on how the persistence of the state can be improved :)

About the stuck operation, I think I know where the problem is. When I tested with a fake garage door opener (modified your code to have print statements instead of GPIO), I had the same problem at first. It appears that when you tap the door to close, iOS expects to get an update of the value for the CurrentDoorState characteristic. Until it receives this, it is stuck on "opening...". So to fix this, I added code that sets the CurrentDoorState to whatever necessary.

So my advice would be to check if the `_detected` method is actually called when the door gets closed. What is more, you can comment out _detected and just set the CurrentDoorState at the end of the pulse_door method to see if this unstucks the whole process.

Let me know how it goes and if you need more assistance/find a bug :)

**Comment by fubano** [comment]
Can you post your fake GarageDoorOpener to accessories if the value posting works fine for you? I am unable to get it to work, sometimes it gets stuck at Opening but Close works fine, or vice-versa, each time I pair. I disabled the sensor callback (though it is working fine hardware-wise) and just post back in the _pulse_door function. Just writing CurrentDoorState does not seem to update the ios client, so I thought it is bound to the TargetDoorState and write back the requested value as seen below. I tried delaying the set_value too, but with no success.

def _pulse_door(self, value):
  print("Target door state is: ",value)
  self.CurrentDoorState.set_value(value)

  GPIO.output(self.pulse_pin, GPIO.HIGH)
  time.sleep(1)
  GPIO.output(self.pulse_pin, GPIO.LOW)
  time.sleep(1)
 
  print("TargetDoorState is:",self.TargetDoorState.get_value())
  print("CurrentDoorState is:",self.CurrentDoorState.get_value())

Thanks!

**Comment by ikalchev** [comment]
Here's the whole thing. I tapped at least 20 times and it didn't stuck once.

```python
from pyhap.accessory import Accessory, Category
import pyhap.loader as loader

class GarageDoor(Accessory):
    category = Category.GARAGE_DOOR_OPENER
    
    def __init__(self, *args, pulse_pin=26, sense_pin=22, **kwargs):
        super(GarageDoor, self).__init__(*args, **kwargs)

        self.CurrentDoorState = self.get_service("GarageDoorOpener").get_characteristic("CurrentDoorState")
        self.ObstructionDetected = self.get_service("GarageDoorOpener").get_characteristic("ObstructionDetected")

    def _pulse_door(self, value):
        print("Door state set to ", value)
        self.CurrentDoorState.set_value(value) # Set the state of the door to the target state

    def _set_services(self):
        super(GarageDoor, self)._set_services()

        garage_door_opener_service = loader.get_serv_loader().get("GarageDoorOpener")
        self.add_service(garage_door_opener_service)
        garage_door_opener_service.get_characteristic("TargetDoorState").setter_callback =     self._pulse_door
```

Logs:
```
IK-Pro:HAP-python-clean ivankalchev$ python3 main.py 
INFO:pyhap.accessory_driver:Starting accessory 'MyDoor' on address '192.168.0.19', port '51826'.
INFO:pyhap.hap_server:Got connection with ('192.168.0.28', 55593).
...
INFO:pyhap.hap_server:192.168.0.28 - "GET /characteristics?id=1.8,1.9,1.10 HTTP/1.1" 207 -
INFO:pyhap.hap_server:192.168.0.28 - "PUT /characteristics HTTP/1.1" 207 -
INFO:pyhap.hap_server:192.168.0.28 - "GET /characteristics?id=1.10,1.9,1.8 HTTP/1.1" 207 -
Door state set to  1
INFO:pyhap.hap_server:192.168.0.28 - "PUT /characteristics HTTP/1.1" 207 -
Door state set to  0
INFO:pyhap.hap_server:192.168.0.28 - "PUT /characteristics HTTP/1.1" 207 -
```

You can try adding the following code to accessory_driver.py#AccessoryDriver line 203 (I will add this as a debug log message soon):
```
for client_addr in subscribed_clients.copy():
    print("Sending ", bytedata, " to client ", client_addr)  # << Add this
    pushed = self.http_server.push_event(bytedata, client_addr)
```
`bytedata` should be the update of the current door state that is being sent to the iOS client. Is this showing up?

**Comment by fubano** [comment]
I do get this exception and it seems the 'bytedata' updates do not occur subsequently.  This is using your code above in a fresh clone, with 'main.py' written by me. Perhaps it has something to do with my python or related package versions?

```
python3 main.py 
INFO:pyhap.accessory_driver:Starting accessory 'MyGDO' on address '192.168.20.117', port '51826'.
INFO:pyhap.hap_server:Got connection with ('192.168.20.112', 49589).
...
INFO:pyhap.hap_server:192.168.20.112 - "GET /characteristics?id=1.10,1.8,1.9 HTTP/1.1" 207 -
INFO:pyhap.hap_server:192.168.20.112 - "PUT /characteristics HTTP/1.1" 207 -
INFO:pyhap.hap_server:192.168.20.112 - "GET /characteristics?id=1.10,1.8,1.9 HTTP/1.1" 207 -
Door state set to  1
Sending  b'{"characteristics": [{"iid": 8, "value": 1, "aid": 1}]}'  to client  ('192.168.20.122', 56746)
INFO:pyhap.hap_server:192.168.20.112 - "PUT /characteristics HTTP/1.1" 207 -
Sending  b'{"characteristics": [{"iid": 8, "value": 1, "aid": 1}]}'  to client  ('192.168.20.107', 63946)
Exception in thread Thread-2:
Traceback (most recent call last):
  File "/usr/lib/python3.4/threading.py", line 920, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.4/threading.py", line 868, in run
    self._target(*self._args, **self._kwargs)
  File "/home/pi/happy2/HAP-python/pyhap/accessory_driver.py", line 206, in send_events
    self.event_queue.task_done()
  File "/usr/lib/python3.4/queue.py", line 71, in task_done
    raise ValueError('task_done() called too many times')
ValueError: task_done() called too many times

INFO:pyhap.hap_server:192.168.20.107 - "PUT /characteristics HTTP/1.1" 207 -
Door state set to  0
INFO:pyhap.hap_server:192.168.20.112 - "PUT /characteristics HTTP/1.1" 207 -
Door state set to  1
INFO:pyhap.hap_server:192.168.20.112 - "PUT /characteristics HTTP/1.1" 207 -
Door state set to  0
INFO:pyhap.hap_server:192.168.20.112 - "PUT /characteristics HTTP/1.1" 207 -
```

If I may make a suggestion, the initial install for me required digging for the right packages to make it work, and the crypto packages were a little hard to find because of the naming. It will be helpful to have a known working list of package versions in the README.

**Comment by ikalchev** [comment]
If you do python3 setup.py install everything is installed for you (dependencies included). If you want to know the actual dependencies, have a look at setup.py (it’s short).

This is python 3.4 on pi zero w right? I didn’t get this exception, so my money on platform/version stuff. Let me welcome new year and I will have a deeper look tomorrow. Thanks for exploring this :)

Best,
Ivan

**Comment by fubano** [comment]
I will redo my PI setup from scratch including OS install and report back if it works fine in some hours.

**Comment by ikalchev** [comment]
Don’t bother if it’s too much effort. Let me have a look before that, because I think it’s version to version issue.

**Comment by ikalchev** [comment]
@fubano, issue was in me. task_done was called for each client, while it should have been called once per event - this caused the thread sending the updates to crash (this is also seen from your logs).

The issue should be resolved with the latest change. Could you please confirm?

Thanks for the effort!

**Comment by fubano** [comment]
Confirmed that it works fine for me now!

I am pretty sure that I used the setup.py when I first installed, perhaps it failed and I did not notice causing me to manually install packages. I use the 'lite' version of raspbian, so this may be the reason. Anyway, I tried re-installing the OS with the newest raspbian 'lite' and I needed a few packages for setup.py to install correctly (I think setuptools and python3-dev).

Grateful for all your help, and wish you a happy new year!
